Improvements to threshold watch

alisonelizabeth opened the pull request:
This PR fixes/improves a few things on the threshold watch page:

- Changed watch actions select dropdown to EUI context menu
- UI improvements to threshold watch form
    - Webhook action form fields inline (I think the original suggestion was to make this one text field, e.g., "webhook url", however, since the advanced watch and threshold watch share some of the same logic, this makes it a little more complicated. I think it's something we can maybe revisit in phase II)
    - Added title to EUI expression
    - Made the `indices` form row full width to give the combo box a little more room
    - Fixed misalignment of comparator fields when invalid
- Added validation for negative interval, no time field
- Removed hard-coded string and cleaned up time options logic
- Updated watch_visualization.tsx to use the `use_request` hook; this adds support for error state 

### Screenshots 

<img width="493" alt="Screen Shot 2019-05-20 at 9 15 37 AM" src="https://user-images.githubusercontent.com/5226211/58024361-ffd5ad00-7adf-11e9-95a9-5952f4e2907e.png">

<img width="677" alt="Screen Shot 2019-05-20 at 9 14 30 AM" src="https://user-images.githubusercontent.com/5226211/58024398-13811380-7ae0-11e9-9974-1e4a2137da10.png">

<img width="953" alt="Screen Shot 2019-05-09 at 4 12 12 PM" src="https://user-images.githubusercontent.com/5226211/58024385-0cf29c00-7ae0-11e9-862f-0253e41374fb.png">

<img width="1177" alt="Screen Shot 2019-05-20 at 9 14 18 AM" src="https://user-images.githubusercontent.com/5226211/58024410-1aa82180-7ae0-11e9-858f-6109ac3d6ed5.png">

<img width="242" alt="Screen Shot 2019-05-20 at 9 14 49 AM" src="https://user-images.githubusercontent.com/5226211/58024422-21369900-7ae0-11e9-9de7-bd8c1e47d8bf.png">
<img width="468" alt="Screen Shot 2019-05-20 at 9 14 58 AM" src="https://user-images.githubusercontent.com/5226211/58024423-21369900-7ae0-11e9-8055-c9b7c509dfc9.png">

<img width="1140" alt="Screen Shot 2019-05-20 at 9 16 21 AM" src="https://user-images.githubusercontent.com/5226211/58024429-2693e380-7ae0-11e9-9225-befb365c7027.png">

elasticmachine commented:
Pinging @elastic/es-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14012/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14039/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14074/)

cjcenizal commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/action_fields/webhook_action_fields.tsx:
```diff
@@ -33,100 +41,108 @@ export const WebhookActionFields: React.FunctionComponent<Props> = ({
-          }}
-        />
-      </EuiFormRow>
+      <EuiFlexGroup justifyContent="spaceBetween">
```
What do you think of prepending the port and path fields with a colon and slash to hint to the user that they don't need to supply those?

![image](https://user-images.githubusercontent.com/1238659/58066556-a6509b00-7b3e-11e9-8053-64f5ce6c5331.png)

```tsx
            <EuiFieldNumber
              prepend={(
                <EuiText size="xs">
                  <strong>:</strong>
                </EuiText>
              )}
              fullWidth
              name="port"
              value={port || ''}
              onChange={e => {
                editAction({ key: 'port', value: parseInt(e.target.value, 10) });
              }}
              onBlur={() => {
                if (!port) {
                  editAction({ key: 'port', value: '' });
                }
              }}
            />
```

```tsx
            <EuiFieldText
              prepend={(
                <EuiText size="xs">
                  <strong>/</strong>
                </EuiText>
              )}
              fullWidth
              name="path"
              value={path || ''}
              onChange={e => {
                editAction({ key: 'path', value: e.target.value });
              }}
            />
```

alisonelizabeth replied:
Great idea, thanks! Updated the code.

cjcenizal commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/threshold_watch_action_dropdown.tsx:
```diff
@@ -47,54 +42,75 @@ interface Props {
+              }}
+            >
+              <EuiFlexGroup>
+                <EuiFlexItem grow={false} style={{ alignSelf: 'center' }}>
```
We have to avoid inline styles in order to not violate our CSP. Can you create a custom class instead?

alisonelizabeth replied:
Good catch - fixed!

cjcenizal commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/threshold_watch_edit.tsx:
```diff
@@ -262,8 +279,16 @@ const ThresholdWatchEditUi = ({ intl, pageTitle }: { intl: InjectedIntl; pageTit
+                  }
                   const theFields = await getFields(indices);
                   setFields(theFields);
+                  setWatchProperty('timeFields', theFields);
```
Minor nit: can we use a name that identifies these fields a bit more than "theFields"? e.g. "timeFields"?

alisonelizabeth replied:
Agreed, it is a little confusing. This is used in a couple different places of the code in reference to all ES fields, and then there is some manipulation in areas to filter only time fields. I renamed this to `esFields`, and `timeFields`, where used. Let me know if you think that's clearer.

cjcenizal commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/threshold_watch_edit.tsx:
```diff
@@ -383,10 +384,19 @@ const ThresholdWatchEditUi = ({ intl, pageTitle }: { intl: InjectedIntl; pageTit
         <EuiSpacer />
         {shouldShowThresholdExpression ? (
           <Fragment>
+            <EuiTitle size="s">
```
Can we use EuiHelpText placed below the expression instead of this error between the title and expression? I think this will follow the pattern of how we surface other field-level errors. And I think we can simplify the error message to "Your expression has errors."

![image](https://user-images.githubusercontent.com/1238659/58066805-b6b54580-7b3f-11e9-98ad-9513519f913a.png)

alisonelizabeth replied:
Good suggestion.

I don't think the `EuiExpression` comes with help text out of the box. However, I went ahead and moved the `EuiText` that existed for this below the expression and reduced the font. Let me know what you think.

<img width="630" alt="Screen Shot 2019-05-21 at 8 36 45 AM" src="https://user-images.githubusercontent.com/5226211/58104284-8a86dc80-7bb2-11e9-9fa3-6abf4f40286a.png">

cjcenizal replied:
Looks great to me!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14308/)

alisonelizabeth commented:
@cjcenizal thanks for the feedback!

I believe I addressed all of your other comments, if you can take another look when you get a chance. I also fixed a couple other things I noticed while I was making changes:
- I noticed the chart was a little glitchy with the reloads, so updated the `useEffect` hook to only be called when relevant watch props have changed
- Updated `Save` button to `secondary` color with icon and loading state; also made text dynamic depending on if it's a new watch or editing an existing watch
- Fixed another react warning related to a number input changing from uncontrolled to controlled

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14319/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14421/)

cjcenizal commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/threshold_watch_edit.tsx:
```diff
@@ -716,7 +717,7 @@ const ThresholdWatchEditUi = ({ intl, pageTitle }: { intl: InjectedIntl; pageTit
                                 >
                                   <EuiFieldNumber
-                                    value={watch.threshold[i]}
+                                    value={watch.threshold[i] || ''}
```
If `watch.threshold[i]` is 0 then this will set `value` to the empty string. Is that what we want? Or should we check for `undefined` and `null` instead of just falsiness, e.g. `watch.threshold[i] == null`.

alisonelizabeth replied:
good catch, fixed.

cjcenizal commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/threshold_watch_edit.tsx:
```diff
@@ -655,8 +662,8 @@ const ThresholdWatchEditUi = ({ intl, pageTitle }: { intl: InjectedIntl; pageTit
                         watchThresholdPopoverOpen ||
-                        errors.threshold0.length ||
-                        (errors.threshold1 && errors.threshold1.length)
+                        !!errors.threshold0.length ||
```
Could you also wrap this expression inside of `Boolean` to achieve the same effect? Might be a bit simpler to read.

```js
isActive={Boolean(
  watchThresholdPopoverOpen ||
  errors.threshold0.length ||
  (errors.threshold1 && errors.threshold1.length)
)}
```

alisonelizabeth replied:
agreed. thanks for the suggestion - fixed!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14497/)
